srtcore/core.cpp

```diff
@@ -10824,15 +10828,21 @@ void CUDT::checkRexmitTimer(const steady_clock::time_point& currtime)
     // Otherwise it might still be sent regulary.
     bool retransmit = false;
     const int32_t unsent_seqno = CSeqNo::incseq(m_iSndCurrSeqNo);
+
+    // TODO: [MAX] Replace the condition with
+    // if (getFlightSpan() > 0 && (!is_laterexmit || m_pSndLossList->getLossLength() == 0))
+    // Do we need two retransmission modes? The only thing that matters is if the peers sends NAK periodically.
```

Comment:
These two different "blind retransmission" procedures remain here because they are working differently and according to requirements of particularly file and live mode.

The blind retransmission is a must in case when you have conditional dropping turned off (which is default in file mode, but can be also set in live mode). The "laterexmit" and "fastrexmit" methods differ mainly by "how stubbornly" the unacknowledged packets are "blindly-retransmitted". The reason for that difference is due to these modes' requirements:

* laterexmit - the solution provided originally in UDT - is used as a "last resort", when you have nothing reported as lost (likely due to lost loss reports themselves), but ACK still doesn't move on. If laterexmit isn't executed, then it may lead to a deadlock situation - the receiver still expects packets to be retransmitted to move the ACK forward, while the sender expect ACK to move forward in order to send further packets. But it's not triggered even for several seconds until it has no reason to retransmit anything on request - we can let ourself do it because file mode doesn't have time requirements

* fastrexmit - the solution provided for live mode - is intended to work in two cases:
   * with TLPKTDROP off and NAKREPORT off - by the same reasons as those described above
   * with only NAKREPORT off - as this would be the only tool for "second lost" and "lossreport-lost" packets (that is, it helps preventing double-lost and lossreport-lost packets from being effectively dropped, while they could still be recovered)

Fastrexmit works more "stubbornly" than laterexmit, which is so because live mode has time requirements so you just can't "wait indefinitely until the sender loss list gets empty", but instead you need to retransmit as fast as it seems to make any sense. This mode would be too excessive for file mode because it risks flooding the link, which might extend the time of link recovery from congestion and this way disturb the calculations in the congestion control module.

Reply:
From the code perspective, the only differences are the following three. All in `CUDT::checkRexmitTimer()`.
Probably not all make much sense. The only meaningful difference is to ignore periodic NAK mode in case of LATEREXMIT (file mode). Maybe it is the only one that should stay.

1. Don't detect loss if FASTREXMIT and PERIODIC_NAK

```c++
    if (is_fastrexmit && m_bPeerNakReport)
        return;
```

2. Additional condition `m_pSndLossList->getLossLength() == 0` in case of LATEREXMIT (file mode).

```c++
    if ((is_laterexmit && unsent_seqno != m_iSndLastAck && m_pSndLossList->getLossLength() == 0)
        // TODO: [MAX] Previously was CSeqNo::seqoff(m_iSndLastAck, unsent_seqno) > 0
         || (is_fastrexmit && getFlightSpan() > 0))
```

3. The `TEV_CHT_FASTREXMIT` type of the event, that anyway does not differ from handling `TEV_CHT_REXMIT`.

```c++
    const ECheckTimerStage stage = is_fastrexmit ? TEV_CHT_FASTREXMIT : TEV_CHT_REXMIT;
    updateCC(TEV_CHECKTIMER, stage);
```

Reply:
In terms of low latency in live mode, the second difference is probably also good. 🤔 
"Additional condition m_pSndLossList->getLossLength() == 0 in case of LATEREXMIT (file mode)."